### DISCUSSION TITLE
authentication in prisma + nextjs

### DISCUSSION THREAD

**Discussion by MohamadrezaKazemian** [prisma/prisma#17135] [Q&A]
Hi.
I built a project with prisma and nextjs but I did not made any structure for authentication of my app.
now I want to add authentication and authorization to my app.

for example this is my schema for User (businessOwner) : 
```
model businessOwners {
  id         String       @id @default(uuid())
  fullName   String
  phone      String       @unique
  email      String       @unique
  username   String       @unique
  password   String
  businesses businesses[]
}
```
and this is my login code (I havent done any validation on it yet):

```
import type { NextApiRequest, NextApiResponse } from 'next';
import prisma from '../../../../lib/prisma';

export default async (req: NextApiRequest, res: NextApiResponse) => {
    if (req.method !== 'POST') {
        return res.status(405).json({ message: 'Method not allowed' });
    }
    try {
        const loginData = JSON.parse(req.body);
        const loginBusinessOwner = await prisma.businessOwners.findFirst({
            where: {
                username: loginData.username,
                password: loginData.password
            },
        });
        if (loginBusinessOwner === null) {
            res.status(400).json(loginBusinessOwner);
        } else {
            res.status(200).json(loginBusinessOwner);

        }
    } catch (err) {
        res.status(400).json({ message: `${err}` });
    }
}; 
```
so now I want to add those things to it. but I have no idea. any help?

**Comment by nurul3101** [comment]
Hello @MohamadrezaKazemian 👋 

You might want to use [Next Auth](https://next-auth.js.org/) to add authentication to your Next.js app.
This is a comprehensive [guide](https://vercel.com/guides/nextjs-prisma-postgres) on how you can integrate Next Auth with Prisma.

**Reply by MohamadrezaKazemian** [reply]
thank you for your replay. i already saw this document. this is for nexAuth + github login. 
i need my own logic function. like using JWT

**Reply by nurul3101** [reply]
NextAuth allows authentication using your own custom implementation, for example with username and password, email and password etc - [Reference](https://authjs.dev/guides/providers/credentials).

**Reply by MohamadrezaKazemian** [reply]
I could not find any good doc!!

my question is, how can we track the user auth info? I mean the authorization.
should we use middleware?

**Reply by nurul3101** [reply]
Yes, you can use Middleware to validate the user credentials.

**Reply by MohamadrezaKazemian** [reply]
is there any example that i can learn from?

**Reply by nurul3101** [reply]
There isn't a dedicated example of using Prisma Middleware for doing authorisation, but here is a [thorough example](https://www.prisma.io/docs/concepts/components/prisma-client/middleware/soft-delete-middleware), this implements soft delete, but the code for authorisation would go into prisma.$use block. Additionally if you want to do authorisation on only specific models or queries, then you can add a condition for matching models/queries as shown in the example.

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that haven’t had any recent activity and appear to be outdated.

If this discussion is still important to you or unresolved, we’d love to hear from you! Feel free to reopen it or start a new one with updated details.

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and being part of the community!
